### DISCUSSION TITLE
How does this work with pipelining?

### DISCUSSION THREAD

**Discussion by feelingsonice** [bikeshedder/deadpool#280] [Q&A] [answered]
Sorry if this is a dumb question, trying to understand how this would work. `tokio_postgres` states:

```
Pipelining happens automatically when futures are polled concurrently
```

So am I correct to then assume that pipelining will never naturally happen? Especially in a standard HTTP server? Because `pool.get()` will always return an unused connection hence there'll never be futures polled concurrently on the same connection?

**Comment by bikeshedder** [comment, marked as answer]
You can perform pipelining with connections from `tokio-postgres` regardless if you create the connection manually or get them from the pool.

There are two use cases:

- [x] **One worker issues two (or more) queries at the same time using a single connection.**
  This works just as explained in the `tokio-postgres` documentation:

  ```rust
  async fn pipelined_prepare(
      client: &Client,
  ) -> Result<(Statement, Statement), Error>
  {
      future::try_join(
          client.prepare("SELECT * FROM foo"),
          client.prepare("INSERT INTO bar (id, name) VALUES ($1, $2)")
      ).await
  }
  ```
- [ ] **Two (or more) workers sharing a single connection and using pipelining that way.**
  This requires a different way to think about pooled objects altogether and is very `tokio-postgres` specific as most DB backends don't really support sharing of connections.
  
  There is a tracking issue for that feature: #196
